driver/src/main/java/org/neo4j/driver/internal/metrics/InternalDriverMetrics.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.driver.internal.metrics;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.neo4j.driver.internal.BoltServerAddress;
+import org.neo4j.driver.internal.async.pool.ConnectionPoolImpl;
+import org.neo4j.driver.internal.metrics.ListenerEvent.ConnectionListenerEvent;
+import org.neo4j.driver.internal.metrics.ListenerEvent.PoolListenerEvent;
+import org.neo4j.driver.internal.metrics.spi.ConnectionMetrics;
+import org.neo4j.driver.internal.metrics.spi.ConnectionPoolMetrics;
+import org.neo4j.driver.internal.spi.ConnectionPool;
+import org.neo4j.driver.v1.Config;
+import org.neo4j.driver.v1.exceptions.ClientException;
+
+import static java.lang.String.format;
+
+public class InternalDriverMetrics extends InternalAbstractDriverMetrics
+{
+    private final Map<String,ConnectionPoolMetrics> connectionPoolMetrics;
+    private final Map<String,ConnectionMetrics> connectionMetrics;
+    private final Config config;
+
+    public InternalDriverMetrics( Config config )
+    {
+        Objects.requireNonNull( config );
+        this.config = config;
+        this.connectionPoolMetrics = new ConcurrentHashMap<>();
+        this.connectionMetrics = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void addMetrics( BoltServerAddress serverAddress, ConnectionPoolImpl pool )
+    {
+        addPoolMetrics( serverAddress, pool );
+        addConnectionMetrics( serverAddress );
```

Comment:
Can there be a race when pool metrics are added but connection metrics are not, and reader sees one but not the other?

Reply:
Internally we will not access any of these metrics until this method returns and a connection pool is created. 

Externally, a user could query the metrics between the time slots where pool metrics are added, but not yet connection metrics. But I guess that does not matter too much? As it is a snapshot-ish thing to external users anyway. If they happens read at that time, then they just get half of the whole view. Then next time when they query the metrics, they will see the newly created ones.